### ISSUE TITLE
[Feature] Add logging for certain events

### ISSUE THREAD

**Issue by Zumochi** [evewspace/eve-wspace#147]
It would be nice to be able to log certain events, so users(/admins) can pull stats from the database (or farther down the road, through a stats page on the interface).

Requested user events that could be logged:
* Added signatures
* Visited systems
* Signatures scanned

Right now, it is logged in the database which user last modified a signature, that gets deleted when the signatures get purged, however.

With this, it might also be an idea to refactor the code that sets the first and last visited date (see #107).

**Comment by Maarten28** [comment]
This is at least (partially) already done, it is just not accessible through the UI. To show you: 

![2015-03-25_19-55-15](https://cloud.githubusercontent.com/assets/3999006/6832789/f8abc572-d328-11e4-8125-cbfb580622b5.png)

One thing which needs to be done is show how many signatures were actually adjusted when using the bulk import. With that information you would be able to create some (reliable) statistics of signatures scanned/updated for example.

**Comment by Zumochi** [comment]
Would that be as ~simple~ as adding `map.add_log()` entries in models.py in the `Signature` class?

If so, I'll try implementing some more logging methods tonight.

**Comment by lynxart** [comment]
Also interested in better logging for signatures and POSes!

**Comment by Zumochi** [comment]
I've made a query that counts signatures updated per user to try it out, maybe it's of use to someone before the full logging stack is implemented. You can of course change the action to whichever action you wish :)

`SELECT user.username, COUNT(1) AS log_count FROM Map_maplog log JOIN account_ewsuser user ON user.id = log.user_id WHERE log.action LIKE 'Updated signature%' GROUP BY user.id ORDER BY log_count DESC`

**Comment by lynxart** [comment]
I'll see if I can add a check to verify whether the sigs updated during a bulk import are actually adjusted or not, and include that info in the logging. Maybe something like "Imported 13 signatures for J102339 (ROOT) -- Updated 8."

**Comment by Maarten28** [comment]
@Zumochi I suggest adding it to the views.py, since all other logging it being tracked there as well.

**Comment by marbindrakon** [comment]
@Maarten28 Logging should be moved to models.py if possible. Map/views.py needs a heavy refactor that I'm working on along with the RESTful API.

**Comment by lynxart** [comment]
@marbindrakon I am adding code to views.py (because I am not sure what kind of restructuring would be needed to integrate the logging into models.py) to track whether imported or edited signatures were "really updated" (adjusted) and whether they were completely scanned down (i.e. Info field was added), as well as how far down the chain the scanning happened. 

Example of what I am capturing: 
User admin imported 4 signatures for J111111(ROOT). Adjusted:  2. Newly scanned: 1. Distance from ROOT: 2.

I'd like to see this recorded in the logging. Would it be best to create a new model for this more detailed scan logging, or should we extend the MapLog model?

**Comment by Zumochi** [comment]
@lynxart Firstly, logging should not be in the view, but in a model. _Please_ try to put them there.

Implementing the logging in models.py _shouldn't_ be that hard, as the method is defined there as well [see this](https://github.com/marbindrakon/eve-wspace/blob/develop/evewspace/Map/models.py#L306).
There are also calls present in models.py, to look at for examples (check add_system at models.py, line 353).

I am not sure how your wish can be implemented as each of those things are separate events.


@marbindrakon if you think [this](https://github.com/marbindrakon/eve-wspace/issues/147#issuecomment-86400495) is a good idea, could you help make a new table with logtypes?

So the structure would add a column `system_id`, and the column `action` would be changed to a foreign key referencing an `action_id` (or something, just putting out an idea).

PS: do you still have a channel on IRC? #eve-wspace is pretty dead, minus the github bot popping up every now and then :smile:

**Comment by lynxart** [comment]
@Zumochi maybe to clarify - I am not implementing any new logging functionality, just making the current logging richer. For example, when bulk importing sigs, I inserted some checks here and there to be able to do things like this:

```Python
# For logging purposes, get the current distance in jumps from the root system
        dist = _get_distance_fromroot(map_system) # new helper function, gets jumps from root
        map_system.map.add_log(
            request.user,
            "Imported %s signatures for %s(%s). Adjusted:  %s. Newly scanned: %s. Distance from ROOT: %s." %
            (k, map_system.system.name, map_system.friendlyname, r, n, dist),
            True
        )
```

The new bits here are the fact that the log now contains a record of whether the signature was modified during import, and whether its imported data was complete (scanned, including info).

In other words, no new logging logic was added, I am just recording some more detailed/relevant log data. I'd like to do something similar for POSes.

**Comment by Zumochi** [comment]
@lynxart that's exactly what I meant, you can put those in the model as well. Please don't put stuff that's supposed to go in the model in the view...
You can extend the logging with the code you just posted in `model.py` as well.

**Comment by lynxart** [comment]
The logging code was _already_ in the view, but OK, I see that it does not really belong there. 

This does mean that I may end up moving the whole bulk_sig_import and related code into the Signature model... because really, that too does not belong in the view, right?

**Comment by Zumochi** [comment]
Partly, yes, but also a good amount of logs are already in the model. Thus it making more sense you also add new ones to the model and not the view :smile:

**Comment by lynxart** [comment]
Thanks for the feedback. I added a log_sig() method to the Signature class, and some smarter signature importing logging for bulk and manual edits in view.py that work with that new method. Sneak preview:

```Python
    def log_sig(self, user, action, map_system):
        """Log the fact that the signature was scanned."""
        map_system.map.add_log(
            user, 
            "%s signature %s in %s (%s), %s jumps out from root system."
            %(action, self.sigid, map_system.system.name, 
              map_system.friendlyname, map_system.distance_from_root()))
```

I now have three actions: Created, Updated and Scanned (for complete info) for the signature logging. Also, distance from root, as you can see, is now a method on MapSystem.

I would love to do the same with POSes, but none of the POS code has a reference to the active map. How could I get that into the view? (I am pretty new to django as you can probably tell).

**Comment by Zumochi** [comment]
So you added them to view.py after all huh :stuck_out_tongue_closed_eyes:

**Comment by lynxart** [comment]
Well... no, I put almost everything in models.py! In fact, I move the _update_from_tsv and _translate_from_client stuff into the model as well. So I'm just calling signature.log_sig from within the signature.update_from_tsv method...

Still figuring out how to pass the map id to the POS update code, so I can log POS updates using the existing logging functionality.

**Comment by lynxart** [comment]
Added POS logging as well.

Pull request https://github.com/marbindrakon/eve-wspace/pull/165 posted.

**Comment by Zumochi** [comment]
I guess this can be closed... Better late than never :)
